src/snowcli/cli/stage/diff.py

```diff
@@ -0,0 +1,151 @@
+import re
+import hashlib
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import List, Dict
+from snowcli.output.types import ObjectResult
+
+from snowflake.connector.cursor import SnowflakeCursor
+from .manager import StageManager
+
+MD5SUM_REGEX = r"^[A-Fa-f0-9]{32}$"
+CHUNK_SIZE_BYTES = 8192
+
+
+@dataclass
+class DiffResult:
+    """
+    Each collection is a list of relative paths, either from
+    the stage root or from the root of the compared local directory.
+    """
+
+    identical: List[str] = field(default_factory=list)
+    "Files with matching md5sums"
+
+    different: List[str] = field(default_factory=list)
+    "Files that may be different between the stage and the local directory"
+
+    only_local: List[str] = field(default_factory=list)
+    "Files that only exist in the local directory"
+
+    only_on_stage: List[str] = field(default_factory=list)
+    "Files that only exist on the stage"
+
+    def display_pretty_diff(self):
```

Comment:
I would avoid names suggesting anything about output format.
IMO logic should only produce a result and should not be aware how it will be formatted.